### PR TITLE
Add/remove nested fields when enabling/disabling unfolded JSON

### PR DESCRIPTION
**Pull request by calherries**
This PR resolves Milestone 2 of https://github.com/metabase/metabase/issues/28667: "Have JSON fields sync automatically when JSON unfolding is toggled on/off". 

Previously, you would have to sync the database to see the effect of toggling on/off JSON unfolding for a field. That's not incredibly intuitive. This PR changes that, so if you toggle off JSON unfolding, the nested fields will be removed (soft deleted) from the table's fields immediately. And if you toggle on JSON unfolding again, the nested fields will be added to the list of active fields, either by activating previously inactive fields, or triggering a sync of the fields in the table.

See the updated [tests](https://github.com/metabase/metabase/blob/40ae0869d4502440864378904cf8e13060c2325d/test/metabase/api/field_test.clj#L732-L781) to understand the impact of the change - syncing the database is now not necessary to have the JSON fields update.

- - -
This change is [Reviewable](https://reviewable.io/reviews/metabase/metabase/29227)

### CODE REVIEW

File: frontend/src/metabase/admin/datamodel/containers/FieldApp.jsx [hunk 1]
```diff
@@ -310,8 +310,7 @@ const FieldGeneralPane = ({
           <Select
             className="inline-block"
             value={
-              field.json_unfolding ||
-              (database.details["json-unfolding"] ?? true)
+              field.json_unfolding ?? database.details["json-unfolding"] ?? true
```
Comment (calherries) [inline frontend/src/metabase/admin/datamodel/containers/FieldApp.jsx:313]: This is in the [unfold-json-by-field](https://github.com/metabase/metabase/tree/unfold-json-by-field) branch, so I'm not sure why it's appearing here in the github UI

File: src/metabase/api/field.clj [hunk 2]
```diff
@@ -117,17 +136,17 @@
    settings           (s/maybe su/Map)
    nfc_path           (s/maybe [su/NonBlankString])
    json_unfolding     (s/maybe s/Bool)}
-  (let [field              (hydrate (api/write-check Field id) :dimensions)
-        new-semantic-type  (keyword (get body :semantic_type (:semantic_type field)))
+  (let [old-field         (hydrate (api/write-check Field id) :dimensions)
```
Comment (calherries) [inline src/metabase/api/field.clj:139]: Renamed `field -> old-field` for clarity

File: src/metabase/api/field.clj [hunk 3]
```diff
@@ -137,23 +156,21 @@
     ;; everything checks out, now update the field
     (api/check-500
      (db/transaction
-       (and
```
Comment (calherries) [inline src/metabase/api/field.clj:140]: This `and` was unnecessary.

File: test/metabase/api/field_test.clj [hunk 4]
```diff
@@ -770,11 +768,12 @@
                 (is (false? (:json_unfolding field))))
               (testing "nested fields are not present since json unfolding is disabled by default"
                 (is (empty? (nested-fields))))
-              (testing "nested fields are removed when json unfolding is disabled"
+              (testing "nested fields aren't added when json unfolding is enabled, because they never existed"
                 (enable-json-unfolding! true)
+                (is (empty? (nested-fields))))
+              (testing "new nested fields are only added when sync is run again"
                 (sync/sync-database! database)
```
Comment (calherries) [inline test/metabase/api/field_test.clj:775]: This sync is the only time where the manual sync is needed - when the nested JSON fields never existed in the first place.

File: test/metabase/api/field_test.clj [hunk 5]
```diff
@@ -744,11 +745,9 @@
           (is (seq (nested-fields))))
         (testing "nested fields are removed when json unfolding is disabled"
           (enable-json-unfolding! false)
-          (sync/sync-database! db)
```
Comment (calherries) [inline test/metabase/api/field_test.clj:747]: Here, sync is not necessary because the fields are automatically updated to have active=false.

File: test/metabase/api/field_test.clj [hunk 6]
```diff
@@ -744,11 +745,9 @@
...
           (is (empty? (nested-fields))))
         (testing "nested fields are added when json unfolding is enabled again"
           (enable-json-unfolding! true)
-          (sync/sync-database! db)
```
Comment (calherries) [inline test/metabase/api/field_test.clj:751]: Here, sync is not necessary because the fields are automatically updated to have active=true.

File: test/metabase/api/field_test.clj [hunk 7]
```diff
@@ -771,11 +770,11 @@
                 (is (false? (:json_unfolding field))))
               (testing "nested fields are not present since json unfolding is disabled by default"
                 (is (empty? (nested-fields))))
-              (testing "nested fields are removed when json unfolding is disabled"
+              (testing "nested fields are added when json unfolding is enabled"
                 (enable-json-unfolding! true)
-                (sync/sync-database! database)
```
Comment (calherries) [inline test/metabase/api/field_test.clj:776]: Here, manual sync is not necessary because the fields are synced automatically.

File: src/metabase/api/field.clj [hunk 8]
```diff
@@ -147,6 +171,7 @@
                                       :present #{:caveats :description :fk_target_field_id :points_of_interest :semantic_type :visibility_type
                                                  :coercion_strategy :effective_type :has_field_values :nfc_path :json_unfolding}
                                       :non-nil #{:display_name :settings}))))
+    (update-nested-fields-on-json-unfolding-change! field json_unfolding)
```
Comment (qnkhuat) [inline src/metabase/api/field.clj:174]: wrap this with a `(when (some? json_unfolding)` ?

File: src/metabase/api/field.clj [hunk 9]
```diff
@@ -98,6 +98,30 @@
                (not (internal-remapping-allowed? base-type new-semantic-type)))
       (t2/delete! Dimension :id old-dim-id))))
 
+(defn- update-nested-fields-on-json-unfolding-change!
+  "If JSON unfolding was enabled for a JSON field, it activates previously synced nested fields from the JSON field.
+   If JSON unfolding was disabled for that field, it inactivates the nested fields from the JSON field.
+   Returns nil."
+  [old-field json_unfolding]
```
Comment (qnkhuat) [inline src/metabase/api/field.clj:105]:
```suggestion
  [old-field json-unfolding]
```

Comment (qnkhuat) [reply src/metabase/api/field.clj:105]: this cond is quite hard to follow IMO, I'd rather have a top-level `(when (not= old-value new-value))` then do the conditional update based on the value of the new value.

**Review by qnkhuat — CHANGES_REQUESTED** [review]
Do we have tests that say: toggle json unfolding via API, then check if all json fields of that field are activated/inactivated?

please point me to it if there is but I can't find it here.

**Comment by calherries** [comment]
@qnkhuat Yes, that's what `json-unfolding-default-false-test` and `json-unfolding-default-true-test` are for. They've been updated so that the manual syncs are removed.

File: src/metabase/api/field.clj (hunk 8 above)
Comment (calherries) [reply src/metabase/api/field.clj]: Updated and tested https://github.com/metabase/metabase/commit/dee4122458c0d60a1cffefd3f6c4d21b2e1f0561

**Review by qnkhuat — APPROVED** [review]
:shipit:

File: src/metabase/api/field.clj [hunk 10]
```diff
@@ -98,6 +98,29 @@
...
+   If JSON unfolding was disabled for that field, it inactivates the nested fields from the JSON field.
+   Returns nil."
+  [old-field new-json-unfolding]
+  (when (not= new-json-unfolding (:json_unfolding old-field))
+    (if new-json-unfolding
+      (let [update-result (t2/update! Field
+                                      :table_id (:table_id old-field)
+                                      :nfc_path [:like (str "[\"" (:name old-field) "\",%]")]
+                                      {:active true})]
+        (when (zero? update-result)
```
Comment (qnkhuat) [inline src/metabase/api/field.clj:112]: can we have test for this case?

**Review by qnkhuat — APPROVED** [review]
🚢

File: test/metabase/api/field_test.clj [hunk 11]
```diff
@@ -783,6 +783,8 @@
                   (set-json-unfolding-for-field! true)
                   (is (seq (nested-fields))))))))))))
 
+(mt/set-test-drivers! [:postgres])
```
Comment (qnkhuat) [inline test/metabase/api/field_test.clj:786]:
```suggestion
```

**Review by dpsutton — APPROVED** [review]
